source/compilation/Compilation.cpp

```diff
@@ -458,14 +458,23 @@ void Compilation::createDefinition(const Scope& scope, LookupLocation location,
     // Record that the given scope contains this definition. If the scope is a compilation unit, add
     // it to the root scope instead so that lookups from other compilation units will find it.
     auto targetScope = scope.asSymbol().kind == SymbolKind::CompilationUnit ? root.get() : &scope;
+    bool redefined = false;
     std::tuple key(def->name, targetScope);
     if (auto it = definitionMap.find(key); it != definitionMap.end()) {
         reportRedefinition(scope, *def, *it->second);
-        return;
+        if (options.allowRedefinition) {
```

Comment:
This should probably only allow redefinition at the top level? i.e. not allow redefining nested modules.

Reply:
I removed any references to `topDefinitions`. It was added at an early stage due to unexplained failures on `ASSERT(!topDef);` , but it is not needed now.

> This should probably only allow redefinition at the top level? i.e. not allow redefining nested modules.

What exactly do you imply by the term "nested modules"? Do you mean any module which is not top level? Is there any special issue with those?

Th case I'm trying to resolve involves reading two non-identical copies of library files, definitely not top level modules. Each (large) library file contains a different set of dozens of small modules, some are common between the 2 files, but some are unique.

Reply:
A nested module is a module definition inside another module. Like so:
```
module m;
  module n;
  endmodule
endmodule
```
Such a module can only be instantiated within its parent definition, and is automatically instantiated once if not explicitly instantiated there. I don't think your change should apply to duplicate nested modules, since nobody would duplicate the definitions on purpose in the same file.

Reply:
Oh, I totally forgot about this SV thing, and I even didn't know it can be automatically instantiated.
I agree that this is irrelevant because in the example above, the scope of `n` is purely within `m`, and it cannot be defined outside `m`.
The only use for this for this feature that comes into my mind is when you incorporate multiple IP's in a SoC where each IP might contain modules with the same name, but of a different revision. you can use this to wrap it:

```
module external_IP1 (/* add ports here, matching IP_file top level */);
`include "IP1_file.v"
`include "IP1/module.v"
IP1 IP1_inst(.*); // connect IP to outside ports
endmodule
```
Now `module.v` will not be visible outside, and you can use a multiple versions of it .

Reply:
> This should probably only allow redefinition at the top level? i.e. not allow redefining nested modules.

Coming back to this again after  a fresh look at the code, a bit more knowledge about implicit creation of std::map entries when a key is not found, and generally reading the relevant sections of Compilation.cpp and Compilation.h over and over:

As far as Iunderstand, the code in its current form will also work for nested modules, not that it is useful or that I've even tested it, but I think that:
```
module x;
  module y;
  endmodule
  module y;
  endmodule
endmodule
```

Would behave this way:

* For module x, it would create an entry using `definitionMap.insert_or_emplace()` in line 468.
* For module y, it would create an entry using `definitionMap.insert_or_emplace()` in line 468. Since `targetScope` is not root, it would create a `topDefinition` entry by setting its bool flag to `true` (line 485). Looking at `topDefinitions[result->name].second = true;` indicates will have no legal `Definition` pointer, since there is no top level module y.
* For the 2nd module y, it would replace the old entry using `definitionMap.insert_or_emplace()` in line 468, as well as setting `redefined` to `true`. Since `targetScope` is not root, it would overwrite the old `topDefinition` entry, writing the 2nd tuple element to `true` again.

Reply:
Aaaand indeed it also triggers for nested modules (in a test named "Nested Modules", that now fails because it checks for the old diag code) - fixed ofcourse.